File: src/01-easy/happy-numbers/solution.cs
using System;
using System.Collections.Generic;
using System.Linq;
class Solution {
    static int Dss(int n){int s=0;while(n>0){int d=n%10;s+=d*d;n/=10;}return s;}
    static void Main(){
        int n=int.Parse(Console.ReadLine());
        while(n-->0){
            string s=Console.ReadLine().Trim();
            int x=s.Sum(c=>(c-'0')*(c-'0'));
            var seen=new HashSet<int>();
            while(x!=1&&seen.Add(x)) x=Dss(x);
            Console.WriteLine($"{s} {(x==1?":)":":(")}" );
        }
    }
}

File: src/01-easy/happy-numbers/solution.rs
use std::io;
use std::collections::HashSet;
fn dss(mut n:i64)->i64{let mut s=0;while n>0{let d=n%10;s+=d*d;n/=10;}s}
fn main(){
    let mut buf=String::new();
    io::stdin().read_line(&mut buf).unwrap();
    let n:usize=buf.trim().parse().unwrap();
    for _ in 0..n{
        let mut s=String::new();
        io::stdin().read_line(&mut s).unwrap();
        let s=s.trim();
        let mut x:i64=s.chars().map(|c|{let d=(c as i64)-48;d*d}).sum();
        let mut seen=HashSet::new();
        while x!=1&&seen.insert(x){x=dss(x);}
        println!("{} {}",s,if x==1{":)"}else{":("}); 
    }
}
